### DISCUSSION TITLE
Login with Device not working on Firefox

### DISCUSSION THREAD

**Discussion by c-rowe** [dani-garcia/vaultwarden#4054] [Q&A]
Hi,  I've updated to the latest version of Vaultwarden and I'm seeing some inconsistency between Chrome and Firefox on the same extension version.  Specifically how login with device works (or doesn't)

Firefox:
![image](https://github.com/dani-garcia/vaultwarden/assets/47050098/5de06c23-f7a9-42b4-a650-6f146914f9ce)

Chrome:
![image](https://github.com/dani-garcia/vaultwarden/assets/47050098/4c14e0bf-a0b7-4654-9e3b-3c4a657b18e9)

I believe that these are the latest for both.  

1) Passkeys work fine for both. Thanks for that, I like it!
2) Login with device works fine in Chrome.
3) Login with device does not work in Firefox.  I get the usual prompt, approve it, etc.  But then it asks for my Fido Key anyway.  After I do auth with the key, it then turns around and asks me for the Master Password again anyway.

4) Tiny niggle: Why do both extensions claim the server version is 2023.9.1 when Vaultwarden Diagnostic shows 2023.10.0?

![image](https://github.com/dani-garcia/vaultwarden/assets/47050098/fcfc801e-d144-4f3d-95c8-7eba6fb81c88)

Otherwise, awesome work guys I'm loving the new features!

Thanks,
-Cameron

**Comment by c-rowe** [comment]
This is what happens when you take screenshots after you write the copy.  I do notice that Chrome has updated to 2023.10.2, but I saw the same behaviour while they were both on 2023.10.1

**Comment by BlackDex** [comment]
> 3\. Login with device does not work in Firefox.  I get the usual prompt, approve it, etc.  But then it asks for my Fido Key anyway.  After I do auth with the key, it then turns around and asks me for the Master Password again anyway.

It works fine for me. You need to remember the 2FA for that browser, after that login with device should work just fine.

> 4\. Tiny niggle: Why do both extensions claim the server version is 2023.9.1 when Vaultwarden Diagnostic shows 2023.10.0?

That is intentional. We have this hardcoded because there currently is only a check by all the clients if the version is equal to or larger then 2023.9.1. we do not know if version 2023.10.x has other features which we do not yet support but could be triggered if the clients detect this. That is why we keep the version on the lowest know supported version.

**Comment by c-rowe** [comment]
> > 3. Login with device does not work in Firefox.  I get the usual prompt, approve it, etc.  But then it asks for my Fido Key anyway.  After I do auth with the key, it then turns around and asks me for the Master Password again anyway.
> 
> It works fine for me. You need to remember the 2FA for that browser, after that login with device should work just fine.

Yeah, I've tried that.  I've saved the 2FA using FIDO, and I've saved the 2FA using TOTP instead.  No luck yet.  After work tonight I'll spin up an entirely clean Firefox install / clean extension install and see if there is something in my Firefox profile that is just unhappy.  Thanks.

> > 4. Tiny niggle: Why do both extensions claim the server version is 2023.9.1 when Vaultwarden Diagnostic shows 2023.10.0?
> 
> That is intentional. We have this hardcoded because there currently is only a check by all the clients if the version is equal to or larger then 2023.9.1. we do not know if version 2023.10.x has other features which we do not yet support but could be triggered if the clients detect this. That is why we keep the version on the lowest know supported version.

Okay, that makes from a code feature safety POV.  Thanks for the explanation.

**Reply by c-rowe** [reply]
Well, I tried it on a clean install of Firefox.  No luck, still the same behaviour.  

So then on a hunch, I tried it with my bitwarden.com account instead of my personal Vaultwarden.  Exact same behaviour.  So, something is amiss with my Firefox (Latest standard from Snap on Ubuntu 22.04 LTS) but it is consistent between Vaultwarden and Bitwarden.  And at the same time my Chrome install on the same machine works fine for both.  

Everyone else in my family is using Chrome or Chrome based browsers, so this is only biting me personally.  Since this isn't exactly an important feature to me, I think I'll leave it there and just check again in a couple months (after a few updates of Firefox and the Add-on)

**Comment by ahmetipkin** [comment]
It's not the 2FA -- 2FA works; but the 'Login with Device'  on the first page is what @c-rowe is mentioning and yes; it's the same for me in Firefox. In Google; it's there but doesn't work with Vaultwarden -- works with original Bitwarden server; though.

**Reply by BlackDex** [reply]
First, the `Login With Device` is only visible after you first logged-in normally. The device needs to be known as a trusted device which has been used to login using a password before.

Second, Websocket's need to work perfectly for this feature to work correctly, If there is anything wrong with your setup and websocket connections do not work, this feature will not work.

Check the `/admin/diagnostics` page to see if there are any issues.

**Reply by ahmetipkin** [reply]
Thanks for that - I checked and all is OK for me; ~~but I still cannot see the 'Passkey login' in the vaultwarden's settings at all (do I need the 'testing' image for that?) It's an already authenticated computer; btw.~~ Confused with another function :)
